### PR TITLE
Fix cong implementation to be properly random and not just cycling.

### PR DESCRIPTION
**Pull request by gbaraldi**
This was found by @IanButterworth. It unfortunately has a small performance regression due to actually using all the rng bits

### CODE REVIEW

File: src/julia_internal.h [hunk 1]
```diff
@@ -1306,20 +1306,36 @@ JL_DLLEXPORT size_t jl_maxrss(void);
...
 // for a small amount of thread-local randomness
 
+//TODO: utilize https://github.com/openssl/openssl/blob/master/crypto/rand/rand_uniform.c#L13-L99
+// for better performance, it does however require making users expect a 32bit random number.
+
 STATIC_INLINE uint64_t cong(uint64_t max, uint64_t *seed) JL_NOTSAFEPOINT
 {
-    if (max == 0)
+    if (max < 2)
         return 0;
```
[the remark this answers is not in the document]
Comment (gbaraldi) [reply src/julia_internal.h:1315]: The old rng is also broken here, which is fun. But basically the issue is that it's a [0, n) interval. So to get 0 and 1 you pass in 2.

[the remark this answers is not in the document]
Comment (gbaraldi) [reply src/julia_internal.h:1315]: Though I'll see if anyone that's using this is actually checking that.

**Comment by gbaraldi** [comment]
@d-netto could you check the change I made to the GC code there. It looked to me that it was assuming that cong return 0 to max in a closed interval and that's not exactly correct.

**Comment by d-netto** [comment]
What matters here is that we want to generate a random number in 0, 1, ... markthreads - 1.

It sees like the previous code was generating random numbers in [0, max], so that's why we have the markthreads - 1 in there.

If the new implementation is generating random numbers kn [0, max) then your change is fine.

**Comment by gbaraldi** [comment]
The old code was also open :)

**Comment by d-netto** [comment]
What about the:

do {...} while (x > max)?

**Comment by d-netto** [comment]
Ah OK, there is a --max in there.

**Comment by gbaraldi** [comment]
The first line in the old implementation does max--. I think when I ported it from blogpost it came from it was there and wrong.

**Comment by IanButterworth** [comment]
I believe we should backport this as the consequences of the scheduler being non-random seems worse than the scheduler being a little slower?

**Comment by oscardssmith** [comment]
Do we have actual examples where the non-randomness causes things to break?

**Comment by IanButterworth** [comment]
Something along these lines? Maybe there's more to attribute the difference to though
```julia
% julia +1.10.5 -t6 
julia> foo() = @sync for _ in 1:Threads.nthreads()
           Threads.@spawn for _ in 1:100000
               yield()
           end
       end
foo (generic function with 1 method)

julia> using Chairmarks

julia> @b foo()
466.335 ms (33 allocs: 3.234 KiB, without a warmup)
```
```julia
% julia +nightly -t6 
julia> versioninfo()
Julia Version 1.12.0-DEV.1122
Commit 1e217279439 (2024-08-30 13:37 UTC)
...

julia> @b foo()
408.224 ms (33 allocs: 2.016 KiB, without a warmup)
```

**Comment by IanButterworth** [comment]
Not clear why the backport labels were removed here.

**Comment by IanButterworth** [comment]
Also, the backport PR commits including references to PRs that weren't included makes it much harder to figure out whether something was actually backported. @KristofferC

**Pull request merged**